clippy2021.md

```diff
@@ -0,0 +1,178 @@
+hitting beta and ultimately stable. This and the fact that fewer people use
+nightly Rust nowadays makes it more probable that a lint with many FPs hits
+stable. This leads to annoyed users, that will disable these new lints in the
+best case and to more annoyed users, that will stop using Clippy in the worst.
+We should think of a process to prevent this from happening. This will also be
+addressed later in this document.
+
+[#6429](https://github.com/rust-lang/rust-clippy/issues/6429)
+
+#### Clippy book
+
+We should consider writing a Clippy book, like other projects have. This will
+then be the central place where everything you need to know to use Clippy is
+collected in a structured way.
+
+[#6011](https://github.com/rust-lang/rust-clippy/issues/6011)
+
+### Reliability
+
+We should make sure to make Clippy more reliable and stable for our users
+
+#### FP rate
+
+In the worst case, new lints are only available in nightly for 2 weeks, before
+hitting beta and ultimately stable. This and the fact that fewer people use
+nightly Rust nowadays makes it more more probably that a lint with many FPs hits
+stable. This leads to annoyed users, that will disable these new lints in the
+best case and to more annoyed users, that will stop using Clippy in the worst.
+We should think of a process to prevent this from happening. This will also be
+addressed later in this issue.
+
+[#6429](https://github.com/rust-lang/rust-clippy/issues/6429)
+
+## Internal
+
+(The end of) 2020 has shown, that Clippy has to think about the available
+resources, especially regarding management and maintenance of the project. This
+section address issues affecting team members and contributors.
+
+### Management
+
+In 2020 we got over 1000 open issues with now between 25-35 open PRs. This is
+mostly because of the lack of resources in the team. According to the planned
+roadmap of the Rust project, we should improve our processes and look for more
+eyes and hands to help reviewing and triaging PRs and issues.
+
+#### Clear expectations for team members
+
+We will have to write this out according to the planned Rust 2021 roadmap
+anyway. This should not put more pressure on the team members, but rather help
+them and interested folks to know what our expectations are.
+
+#### More maintainer
+
+More people means less work for each individual. We should work out a process
+how onboarding of new folks can work and look out for people that are interested
+joining and taking up responsibility.
+
+#### (Bi-)weekly meetings
+
+Other teams have regular meetings. Clippy is big enough, that it might be worth
+it to also do them. Especially if we want/get more people joining the team, this
+can be important for sync-ups. I for example can't keep track of all the new
+lints and features that are implemented in Clippy, because I stopped looking at
+every PR a while ago. And no one should have to do this to be up-to-date.
+
```

Comment:
Meetings might help. I'm not sure: i like asynchronous communication, and I feel like clippy is an ideal project for that to work since 90% of the stuff that happens is work on individual lints, and I largely feel it's _fine_ to not have a finger on the pulse for those. But as is glaringly obvious from this document, we have a _ton_ of large, high level projects we can do, and more synchronous communication would probably help them happen!

Writing this comment here not as a disagreement, really, just to create a point for discussion.

Reply:
> Writing this comment here not as a disagreement, really, just to create a point for discussion.

That's exactly what I wanted to achieve with this. What I wrote down is my perspective on things, where I would like to hear the opinion from others. 

---

As for the meetings: I think short meetings will help if we want to tackle the bigger tasks in this document. 

Additionally, we could do those for example bi-weekly before a Rust sync. That way people can raise concerns about things that should not be synced as-is or other things that are on their mind. Even when the meeting just consists of "no concerns" & "no updates from bigger tasks" and we end it after 2 minutes, that's fine. This might be especially helpful if we decide to grow the team, to get everyone on the same page.

Also I don't really get to look at issues anymore. So meetings like this can be helpful if someone wants to point out something we should discuss/address as a team instead of individuals. Or people that currently "only" do triage can bring up stuff where they see a trend in the issues.

=> If no one is against having regular meetings, I would like to try them out. If we find that they're not helpful, we can stop them again.

Reply:
One thing I know for release team has definitely been true is that *having* a standing meeting, even if it's usually just 5 minutes or so, that you don't cancel even with an empty agenda can feel quite odd (what are we here for) but I personally find pretty useful. That meeting is weekly and scheduled for 30 minutes on Zulip, but it does mean that if we have some problem or issue, we can discuss at a known time. I find it can be helpful to just have a standing synchronous venue for some quick discussion.

(FWIW this sort of "do I run meetings? how do I run meetings?" work is part of the goals for the 2021 Rust roadmap charter work -- seeing the patterns across the project I think is going to be pretty helpful :)

Reply:
> FWIW this sort of "do I run meetings? how do I run meetings?" work is part of the goals for the 2021 Rust roadmap charter work

I know, that's where I got the idea for writing (some of) this. :)

Reply:
> Hm, now I wonder if that's a result of asynchronous communication causing those tasks to be worked on more frequently than tasks that would require meetings first?

Yeah, that's kinda what I'm saying, for most stuff we don't need meetings, but we're large enough that it's now becoming easy to ignore the big ticket items. And we clearly have a lot of them!

Biweekly before a sync is actually kinda brilliant. The devtools meetings are monthly, text-based (i.e., on Discord, could use Zulip), and are relatively low effort.

On triage: it's worth wondering if we can loop in the existing triage team to help. If we wrote a TRIAGING doc that lets others triage things here that would be cool.